accepted/rfc0000-scope-scheduling.md

```diff
@@ -0,0 +1,325 @@
+[`spawn`] functions respectively, except that they ensure **per-thread
+FIFO** ordering. The `ScopeFifo` struct is analogous to existing
+[`Scope`] struct -- it permits one to spawn new tasks that will
+execute before the `scope_fifo` function returns.
+
+[`scope`]: https://docs.rs/rayon/1.0.3/rayon/fn.scope.html
+[`spawn`]: https://docs.rs/rayon/1.0.3/rayon/fn.spawn.html
+[`Scope`]: https://docs.rs/rayon/1.0.3/rayon/struct.Scope.html
+
+## Deprecations
+
+The `breadth_first` flag on thread-pools is **deprecated** but (for
+now) retains its current behavior. In some future rayon-core release,
+it may become a no-op, so users are encouraged to migrate to use
+`scope_fifo` or `spawn_fifo` instead.
+
+# Implementation notes
+
+## Implementing `scope_fifo`
+
+Rayon's core thread pool operates on the traditional work-stealing
+mechanism, where each worker thread has its own deque. New tasks are
+pushed onto the back of the deque, and to obtain a local task, the
+thread pops from the back of the deque. When stealing, jobs are taken
+from the front of the deque. So how can we extend this to accommodate
+the new scheduling modes? 
+
+In addition, we have the goal that nested scopes using these modes
+should "compose" nicely with one another (and with the `join`
+operation)[^global]. So, for example, if we have a thread nesting scopes like:
+
+[^global]: The existing "global FIFO mode" fails miserably on this criteria, which is a partial motivator for this RFC.
+
+- a per-thread LIFO scope S1 that contains a
+  - per-thread FIFO scope S2 that contains
+    - a join(A, B) of tasks A and B
+    
+then we should execute:
+
+- first, the tasks from the join (in reverse order, so B and then A)
+- then, the tasks from S2, in the order that they were created
+- then, the tasks from S1, in the reverse order from which they were created.
+
+Implementing **Per-thread LIFO** scheduling is therefore very
+easy. Each new job pushed onto the stack is simply pushed directly
+onto the back of the deque (if the `breadth_first` flag is true, then
+new tasks are pushed onto the front of the deque). Worker threads pop
+local tasks from the back of the deque but steal from the front.
+
+Implementing **Per-thread FIFO** requires a certain amount of
+indirection. The scope creates N FIFOs, one per worker thread (as of
+this writing, the number of worker threads is fixed; if we later add
+the ability to grow or shrink the thread-pool dynamically, that will
+make this implementation somewhat more complex). When a new task is
+pushed onto a FIFO scope by the worker with index W, we actually push two items:
+
+- First, we push the task itself onto the FIFO with index W.
+  This task contains the closure that needs to execute.
+- Second, we push an "indirect" task onto the worker's thread-local
+  deque. This task records the worker index W that created it, but
+  does not record the actual closure that needs to execute.
```

Comment:
Minor note - as currently implemented in the PR, the indirect task is actually a handle straight to the thread-specific FIFO.  So executing this doesn't require any further lookup, just pop the FIFO and go.